api/app.go

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2015 Spinpunch, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package api
+
+import (
+	l4g "code.google.com/p/log4go"
+	"fmt"
+	"github.com/gorilla/mux"
+	"github.com/mattermost/platform/model"
+	"github.com/mattermost/platform/utils"
+	"net/http"
+	"net/url"
+)
+
+func InitApp(r *mux.Router) {
+	l4g.Debug("Initializing app api routes")
+
+	sr := r.PathPrefix("/apps").Subrouter()
+
+	sr.Handle("/oauth2/register", ApiUserRequired(registerOAuthApp)).Methods("POST")
+	sr.Handle("/oauth2/allow", ApiUserRequired(allowOAuth)).Methods("GET")
+	sr.Handle("/oauth2/access_token", ApiAppHandler(getAccessToken)).Methods("POST")
+}
+
+func registerOAuthApp(c *Context, w http.ResponseWriter, r *http.Request) {
+	app := model.AppFromJson(r.Body)
+
+	if app == nil {
+		c.SetInvalidParam("registerOAuthApp", "app")
+		return
+	}
+
+	secret := model.NewId()
+
+	app.ClientSecret = secret
+	app.CreatorId = c.Session.UserId
+
+	if result := <-Srv.Store.App().Save(app); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		app = result.Data.(*model.App)
+		app.ClientSecret = secret
+
+		c.LogAudit("client_id=" + app.Id)
+
+		w.Write([]byte(app.ToJson()))
+		return
+	}
+
+}
+
+func allowOAuth(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.LogAudit("attempt")
+
+	w.Header().Set("Content-Type", "application/x-www-form-urlencoded")
+	responseData := map[string]string{}
+
+	responseType := r.URL.Query().Get("response_type")
+	if len(responseType) == 0 {
+		c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Bad response_type")
+		return
+	}
+
+	clientId := r.URL.Query().Get("client_id")
+	if len(clientId) != 26 {
+		c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Bad client_id")
+		return
+	}
+
+	redirectUri := r.URL.Query().Get("redirect_uri")
+	scope := r.URL.Query().Get("scope")
+	state := r.URL.Query().Get("state")
+
+	var app *model.App
+	if result := <-Srv.Store.App().Get(clientId); result.Err != nil {
+		c.Err = model.NewOAuthAppError("allowOAuth", "server_error", "Error accessing the database")
```

Comment:
what happens when clientId is missing?  do we need to give back a forbidden code?